File: backend/app/auth/admin_service.py
from sqlalchemy import select
from sqlalchemy.orm import Session
from backend.app.database.models import User, UserRole

def get_all_users(db: Session) -> list[User]:
    return list(
        db.scalars(
            select(User).order_by(User.id)
        ).all()
    )

def update_user_role(
    db: Session,
    user_id: int,
    new_role: str,
    current_admin_id: int,
) -> User:

    user = db.get(User, user_id)

    if user is None:
        raise ValueError("User not found.")

    allowed_roles = {
        UserRole.USER.value,
        UserRole.ANALYST.value,
        UserRole.ADMIN.value,
    }

    if new_role not in allowed_roles:
        raise ValueError(
            "Invalid role. Allowed roles: USER, ANALYST, ADMIN."
        )

    if user.id == current_admin_id:
        raise ValueError(
            "Admin cannot change their own role."
        )

    user.role = new_role

    db.commit()
    db.refresh(user)

    return user

def update_user_status(
    db: Session,
    user_id: int,
    is_active: bool,
    current_admin_id: int,
) -> User:

    user = db.get(User, user_id)

    if user is None:
        raise ValueError("User not found.")

    if user.id == current_admin_id and not is_active:
        raise ValueError(
            "Admin cannot deactivate their own account."
        )

    user.is_active = is_active

    db.commit()
    db.refresh(user)

    return user

File: spring-backend/src/main/java/com/finguard/springbackend/service/AdminService.java
package com.finguard.springbackend.service;

import com.finguard.springbackend.entity.User;
import com.finguard.springbackend.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AdminService {

    private final UserRepository userRepository;

    public AdminService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public List<User> getAllUsers() {
        return userRepository.findAll()
                .stream()
                .sorted((a, b) -> a.getId().compareTo(b.getId()))
                .toList();
    }

    public User updateUserRole(
            Long userId,
            String newRole,
            Long currentAdminId
    ) {
        User user = userRepository.findById(userId)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "User not found."
                        )
                );

        if (!newRole.equals("USER")
                && !newRole.equals("ANALYST")
                && !newRole.equals("ADMIN")) {
            throw new IllegalArgumentException(
                    "Invalid role. Allowed roles: USER, ANALYST, ADMIN."
            );
        }

        if (user.getId().equals(currentAdminId)) {
            throw new IllegalArgumentException(
                    "Admin cannot change their own role."
            );
        }

        user.setRole(newRole);

        return userRepository.save(user);
    }

    public User updateUserStatus(
            Long userId,
            boolean isActive,
            Long currentAdminId
    ) {
        User user = userRepository.findById(userId)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "User not found."
                        )
                );

        if (user.getId().equals(currentAdminId) && !isActive) {
            throw new IllegalArgumentException(
                    "Admin cannot deactivate their own account."
            );
        }

        user.setActive(isActive);

        return userRepository.save(user);
    }
}